### PR TITLE
Fixed #31494 -- Preserved query strings when following HTTP 307/308 redirects in test client.

### PR DESCRIPTION
**Pull request by lothemar**

### CODE REVIEW

**Comment by Valze** [comment]
Fixing this patch as patch doesn't correctly address the ticket.

Namely, the patch should include the final query string when /foo/ is redirected to /bar/?a=b/

This patch doesn't fix that, but instead maintains the query string from /foo/?a=b/ to /bar/?a=b/

Submitting a fix within the hour.

**Comment by smithdc1** [comment]
Hi @Valze 

As this is a new feature should add to the 3.2 release notes?

**Comment by Valze** [comment]
> 
> 
> Hi @Valze
> 
> As this is a new feature should add to the 3.2 release notes?

Good catch, will do!

File: django/test/client.py [hunk 1]
```diff
@@ -833,6 +833,8 @@ def _handle_redirects(self, response, data='', content_type='', **extra):
                 data = QueryDict(url.query)
                 content_type = None
 
+            if response.request['REQUEST_METHOD'] != 'GET':
+                path = urljoin(path, '?' + url.query)
```
Comment (felixxm) [inline django/test/client.py:837]: As far as I'm concerned this should impact only `307` and `308` redirects, so maybe:
```diff
diff --git a/django/test/client.py b/django/test/client.py
index b26504f762..e4201bead4 100644
--- a/django/test/client.py
+++ b/django/test/client.py
@@ -827,7 +827,10 @@ class Client(ClientMixin, RequestFactory):
 
             if response.status_code in (HTTPStatus.TEMPORARY_REDIRECT, HTTPStatus.PERMANENT_REDIRECT):
                 # Preserve request method post-redirect for 307/308 responses.
-                request_method = getattr(self, response.request['REQUEST_METHOD'].lower())
+                request_method = response.request['REQUEST_METHOD'].lower()
+                if request_method != 'get':
+                    extra['QUERY_STRING'] = url.query
+                request_method = getattr(self, request_method)
             else:
                 request_method = self.get
                 data = QueryDict(url.query)
```

File: docs/releases/3.2.txt [hunk 2]
```diff
@@ -199,6 +199,9 @@ Tests
   now isolated for each test method. Such objects are now required to support
   creating deep copies with :py:func:`copy.deepcopy`. Assigning objects which
   don't support ``deepcopy()`` is deprecated and will be removed in Django 4.1.
+* 307 and 308 redirects made with the :class:`~django.test.Client` will now
```
Comment (felixxm) [inline docs/releases/3.2.txt:202]: Add blank line.

File: tests/test_client/tests.py [hunk 3]
```diff
@@ -284,6 +284,18 @@ def test_follow_307_and_308_redirect(self):
                 self.assertEqual(response.request['PATH_INFO'], '/post_view/')
                 self.assertEqual(response.request['REQUEST_METHOD'], method.upper())
 
+    def test_follow_307_and_308_preserves_query_string(self):
+        methods = ('get', 'post', 'head', 'options', 'put', 'patch', 'delete', 'trace')
```
Comment (felixxm) [inline tests/test_client/tests.py:288]: `GET` works without this patch. Extra tests are fine but should be moved to a separate commit.

File: tests/test_client/tests.py [hunk 4]
```diff
@@ -284,6 +284,18 @@ def test_follow_307_and_308_redirect(self):
...
+        codes = (307, 308)
+        for method, code in itertools.product(methods, codes):
+            with self.subTest(method=method, code=code):
+                req_method = getattr(self.client, method)
+                response = req_method('/redirect_query_%s/' % code, data={'value': 'test'}, follow=True)
```
Comment (felixxm) [inline tests/test_client/tests.py:293]: I'm not sure why we pass `data` and build a query string in tests views :thinking: I would simplify this:
```python
def test_follow_307_and_308_no_get_preserves_query_string(self):
    methods = ('post', 'head', 'options', 'put', 'patch', 'delete', 'trace')
    codes = (307, 308)
    for method, code in itertools.product(methods, codes):
        with self.subTest(method=method, code=code):
            req_method = getattr(self.client, method)
            response = req_method('/redirect_query_%s/' % code, follow=True)
            self.assertRedirects(response, '/post_view/?hello=world', status_code=code)
```

and in `views.py`:

```python
def method_saving_307_query_view(request):
    return HttpResponseRedirect('/post_view/?hello=world', status=307)


def method_saving_308_query_view(request):
    return HttpResponseRedirect('/post_view/?hello=world', status=308)
```

Maybe I'm missing sth.

Comment (lothemar) [reply tests/test_client/tests.py:293]: Makes perfect sense, will simplify it.

Still very much a newbie at the test client so learning as I go

File: django/test/client.py (hunk 1 above)
Comment (lothemar) [reply django/test/client.py:837]: Agreed. I originally had it nested within the 307, 308 if block but moved it out

File: tests/test_client/tests.py (hunk 4 above)
Comment (felixxm) [reply tests/test_client/tests.py]: @lothemar I realized that previous assertions were correct. The current tests work even without this patch. I will restore them, sorry.

**Comment by felixxm** [comment]
@lothemar Thanks :+1:

**Comment by felixxm** [comment]
Merged in 7c947f0f5a0539add9adf8c4e14726ea509e13ef and 6425fd3124e50219ba0b6f545d1bf7edfb053a59.
